src/keys/slip10.rs

```diff
@@ -44,100 +99,404 @@ impl Seed {
+        self.0.secret_bytes()
+    }
 
-impl Key {
     pub fn secret_key(&self) -> SecretKey {
-        let mut il = [0; 32];
-        il.copy_from_slice(&self.0[..32]);
-        SecretKey::from_bytes(il)
+        match self.curve() {
+            #[cfg(feature = "ed25519")]
+            Curve::Ed25519 => SecretKey::Ed25519(ed25519::SecretKey::from_bytes(*self.secret_bytes())),
+            #[cfg(feature = "secp256k1")]
+            Curve::Secp256k1 => secp256k1_ecdsa::SecretKey::try_from_bytes(*self.secret_bytes())
+                .map(SecretKey::Secp256k1Ecdsa)
+                .expect("valid extended secret key"),
+        }
     }
 
-    pub fn chain_code(&self) -> ChainCode {
-        let mut ir = [0; 32];
-        ir.copy_from_slice(&self.0[32..]);
-        ir
+    pub fn chain_code(&self) -> &[u8; 32] {
+        self.0.chain_code()
     }
 
-    pub fn child_key(&self, segment: &Segment) -> crate::Result<Key> {
-        if !segment.hardened {
+    pub fn child_key(&self, segment: &Segment) -> crate::Result<Self> {
+        #[cfg(feature = "ed25519")]
+        if self.0.curve == Curve::Ed25519 && !segment.is_hardened() {
             return Err(crate::Error::InvalidArgumentError {
                 alg: "SLIP10",
-                expected: "hardened key",
+                expected: "hardened key index for Ed25519 master secret key",
             });
         }
 
-        let mut data = [0u8; 1 + 32 + 4];
-        data[1..1 + 32].copy_from_slice(&self.0[..32]); // ser256(k_par) = ser256(parse256(il)) = il
-        data[1 + 32..1 + 32 + 4].copy_from_slice(&segment.bs); // ser32(i)
+        Ok(Self(self.0.child_key(segment)))
+    }
 
-        let mut i = [0; 64];
-        HMAC_SHA512(&data, &self.0[32..], &mut i);
+    pub fn derive(&self, chain: &Chain) -> crate::Result<Self> {
+        #[cfg(feature = "ed25519")]
+        if self.0.curve == Curve::Ed25519 && !chain.all_hardened() {
+            return Err(crate::Error::InvalidArgumentError {
+                alg: "SLIP10",
+                expected: "hardened key index for Ed25519 master secret key",
+            });
+        }
+
+        let mut key = self.0.clone();
+        for segment in &chain.0 {
+            key = key.child_key(segment);
+        }
+        Ok(Self(key))
+    }
+
+    pub fn try_into_extended_public_key(&self) -> crate::Result<ExtendedPublicKey> {
+        if self.curve().is_non_hardened_supported() {
+            let mut k = KeyImpl {
+                curve: self.curve(),
+                ext: [0_u8; 65],
+            };
+            k.ext[..33].copy_from_slice(&self.0.calc_public_bytes());
+            k.ext[33..].copy_from_slice(self.chain_code());
+            Ok(ExtendedPublicKey(k))
+        } else {
+            Err(crate::Error::InvalidArgumentError {
+                alg: "SLIP10",
+                expected: "curve supporting non-hardened key derivation",
+            })
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct ExtendedPublicKey(KeyImpl);
```

Comment:
Can you elaborate on what is an extended public key? Compared to a public key.

Reply:
Extended public key contains public key and chain code (and a curve). It allows it to derive child non-hardened public keys. It is the terminology [from the slip](https://github.com/satoshilabs/slips/blob/master/slip-0010.md?plain=1#L115).